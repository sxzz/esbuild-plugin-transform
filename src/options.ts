import { type OnLoadOptions, type Plugin } from 'esbuild'

export interface Options extends Partial<OnLoadOptions> {
  namespace?: string
  plugins?: Plugin[]
}

export type OptionsResolved = Omit<Required<Options>, 'namespace'> & {
  namespace?: undefined | string
}

export function resolveOptions(options: Options): OptionsResolved {
  return {
    filter: /.*/,
    namespace: undefined,
    plugins: [],
    ...options,
  }
}
