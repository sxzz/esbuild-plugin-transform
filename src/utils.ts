import { tmpdir } from 'os'
import fs, { copyFileSync, existsSync, writeFileSync } from 'fs'
import path from 'path'
import { ensureDirSync } from 'fs-extra'

function random() {
  return Math.trunc(Math.random() * 10000)
}

export function backupFile(
  baseDir: string,
  dirName: string,
  filename: string,
  contents?: Uint8Array | string
) {
  let id = random()
  let backupPath: string

  do {
    backupPath = path.resolve(baseDir, `${dirName}${id}`, filename)
    if (existsSync(backupPath)) {
      id++
      continue
    }

    ensureDirSync(path.dirname(backupPath))
    if (!contents) {
      const originalFile = path.resolve(baseDir, filename)
      copyFileSync(originalFile, backupPath, fs.constants.COPYFILE_EXCL)
    } else {
      writeFileSync(backupPath, contents, 'utf-8')
    }
    return backupPath
    // eslint-disable-next-line no-constant-condition
  } while (true)
}

const tempDir = tmpdir()
export function createTempFile(
  filePath: string,
  contents?: string | Uint8Array
) {
  const filename = path.isAbsolute(filePath) ? filePath.slice(1) : filePath
  return backupFile(tempDir, '', filename, contents ?? '')
}
