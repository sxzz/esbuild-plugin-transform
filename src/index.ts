import { existsSync } from 'node:fs'
import path from 'node:path'
import { rm, writeFile } from 'node:fs/promises'
import {
  type OnLoadArgs,
  type OnLoadOptions,
  type OnLoadResult,
  type Plugin,
  type PluginBuild,
} from 'esbuild'
import { type Options, resolveOptions } from './options'
import { backupFile, createTempFile } from './utils'

export const Transform = (userOptions: Options = {}): Plugin => {
  const options = resolveOptions(userOptions)
  const onLoads: Parameters<PluginBuild['onLoad']>[] = []

  return {
    name: 'esbuild-plugin-transform',
    async setup(build) {
      for (const plugin of options.plugins) {
        await plugin.setup({
          ...build,
          onLoad(...args) {
            onLoads.push(args)
          },
        })
      }

      build.onLoad(
        { filter: options.filter, namespace: options.namespace },
        async (args) => {
          const isRealModule = existsSync(args.path)

          let transformedPath: [string, string] | undefined
          let transformed: OnLoadResult = { pluginData: args.pluginData }
          let contents: string | Uint8Array | undefined

          function getTransformedFile() {
            // Returns original path when first plugin
            if (!contents) return args.path

            if (!transformedPath) {
              if (isRealModule) {
                const dirName = path.dirname(args.path)
                const filename = path.basename(args.path)
                transformedPath = backupFile(
                  dirName,
                  'transformed',
                  filename,
                  contents
                )
              } else {
                transformedPath = createTempFile(args.path, contents)
              }
            }
            return transformedPath[1]
          }

          for (const [options, onLoad] of onLoads) {
            if (!filterOnLoad(options, args)) continue

            const result = await onLoad({
              ...args,
              get path() {
                return getTransformedFile()
              },
              pluginData: transformed.pluginData,
            })
            if (!result || typeof result.contents === 'undefined') continue

            transformed = result
            contents = result.contents

            if (transformedPath)
              await writeFile(transformedPath[1], contents, 'utf-8')
          }

          if (transformedPath) {
            await rm(transformedPath[0], {
              recursive: true,
              force: true,
            })
          }

          if (typeof transformed.contents === 'undefined') return
          return transformed
        }
      )
    },
  }
}

function filterOnLoad(options: OnLoadOptions, args: OnLoadArgs) {
  const matchNamespace =
    typeof options.namespace !== 'undefined'
      ? options.namespace === args.namespace
      : true
  return options.filter.test(args.path) && matchNamespace
}
