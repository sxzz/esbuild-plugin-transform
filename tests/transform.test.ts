/* eslint-disable unicorn/prefer-string-replace-all */

import path from 'path'
import { TextDecoder } from 'util'
import { readFile } from 'fs/promises'
import { expect, test } from 'vitest'
import { build } from 'esbuild'
import { Transform } from '../src'
import type { Plugin } from 'esbuild'

const fixtures = path.resolve(__dirname, 'fixtures')

async function getOutput(entry: string, plugins: Plugin[]) {
  const output = await build({
    entryPoints: [entry],
    write: false,
    plugins: [Transform({ plugins })],
  })
  return new TextDecoder().decode(output.outputFiles[0].contents).trim()
}

function PluginWrapper(
  name: string,
  transformer: (contents: string) => Promise<string> | string
): Plugin {
  return {
    name,
    setup: (build) => {
      build.onLoad({ filter: /\.ts$/ }, async ({ path }) => {
        const contents = await readFile(path, 'utf-8')
        return {
          contents: await transformer(contents),
          loader: 'ts',
        }
      })
    },
  }
}

test('transform', async () => {
  const plugin1: Plugin = PluginWrapper('plugin1', (contents) =>
    contents.replace(/foo/g, 'bar')
  )
  const plugin2: Plugin = PluginWrapper('plugin2', (contents) =>
    contents.replace(/bar/g, 'baz')
  )

  const contents = await getOutput(path.resolve(fixtures, 'basic.ts'), [
    plugin1,
    plugin2,
  ])

  expect(contents).toMatchInlineSnapshot(
    `
    "\\"use strict\\";
    const baz = \\"baz\\";"
  `
  )
})
