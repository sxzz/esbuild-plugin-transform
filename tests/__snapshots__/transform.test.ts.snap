// Vitest Snapshot v1

exports[`transform > basic 1`] = `
"\\"use strict\\";
(() => {
  // tests/fixtures/basic.ts
  var expected = \\"expected\\";
})();"
`;

exports[`transform > virtual module 1`] = `
"\\"use strict\\";
(() => {
  // vmodule:/vmodule/foo
  var expected = \\"expected\\";
})();"
`;
